modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/js/components/control-menu/ControlMenu.es.js

```diff
@@ -15,7 +15,7 @@
 import ClayIcon from '@clayui/icon';
 import classNames from 'classnames';
 import React, {useContext, useEffect, useState} from 'react';
-import {createPortal} from 'react-dom';
+import {ReactPortal} from '@liferay/frontend-js-react-web';
```

Comment:
Just wondering here, why create a component instead of

```
import {createPortal} from '@liferay/frontend-js-react-web';
```

I think that's what we tried to do with [`render`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx).

Reply:
I quite like  `<ReactPortal>`, it looks nicer and more readable than `createPortal()`

But, without going in too deep, it seems that we are doing a lot of new stuff in  [ReactPortal](https://github.com/liferay-frontend/liferay-portal/pull/1016/commits/82a72f55be8cd1ff172d3ad8cd677f6d05212ae7). It not just a thin wrapper around `createPortal` that sets `'lfr-tooltip-scope'`. Is it just necessary boilerplate for it to work as a component, or is there more to it?

Reply:
I actually just copied our implementation from [ClayPortal](https://github.com/liferay/clay/blob/master/packages/clay-shared/src/Portal.tsx) with a few small changes (container instead of containerRef).

> Just wondering here, why create a component instead of

Partly because it was just copied clay, but I also think by exposing it as a component we can help direct the mental model of how portals are used. Currently, createPortal in DXP seems to be used as a way to attach components to various nodes around the DOM, similar to instantiating an AUI component. I think is a bad pattern for us and it's going to be better long term for us to see Portals as an intentional break in the DOM flow but within the React App's flow. Right now I think `createPortal` is sort of being abused as a `ReactDOM.render` replacement. Do a look up on createPortal in DXP and follow some of the logic flow, it's worrisome with how we are passing id's in and then manually querying the DOM for a very specific node.


> But, without going in too deep, it seems that we are doing a lot of new stuff in ReactPortal. It not just a thin wrapper around createPortal

@markocikos, the new stuff in ReactPortal is for properly nesting portals. Rather than attaching everything to document.body, we check to see if the Portal is within an existing portal and treats the root node of the top-level portal as the root. This helps to keep us from having numerous empty `<div />` elements at the bottom of the body. We used this in Clay and if you want to read some of the background on it, [check out this issue](https://github.com/liferay/clay/issues/2246). It might not be super necessary right away for DXP, but I think it would be beneficial down the road.


Note, we could expose ClayPortal to be more friendly and consumable by DXP. Right now since ClayPortal is in our internal `@clayui/shared` package, we don't want to encourage its use outside of Clay yet.